sync-core/src/main/java/com/cloudant/http/HttpConnection.java

```diff
@@ -0,0 +1,251 @@
+//  Copyright (c) 2015 IBM Cloudant. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+//  except in compliance with the License. You may obtain a copy of the License at
+//  http://www.apache.org/licenses/LICENSE-2.0
+//  Unless required by applicable law or agreed to in writing, software distributed under the
+//  License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+//  either express or implied. See the License for the specific language governing permissions
+//  and limitations under the License.
+
+package com.cloudant.http;
+
+import com.cloudant.android.Base64OutputStreamFactory;
+
+import org.apache.commons.io.IOUtils;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.HashMap;
+
+/**
+ * Created by tomblench on 23/03/15.
+ */
+
+/**
+ * <p>
+ * A wrapper for <code>HttpURLConnection</code>s.
+ * </p>
+ *
+ * <p>
+ * Provides some convenience methods for making requests and sending/receiving data as streams,
+ * strings, or byte arrays.
+ * </p>
+ *
+ * <p>
+ * Typical usage:
+ * </p>
+ *
+ * <pre>
+ * HttpConnection hc = new HttpConnection("POST", "application/json", new URL("http://somewhere"));
+ * hc.setUsername("user");
+ * hc.setPassword("pass");
+ * hc.requestProperties.put("x-some-header", "some-value");
+ * hc.setRequestBody("{\"hello\": \"world\"});
+ * String result = hc.execute().responseAsString();
+ * // get the underlying HttpURLConnection if you need to do something a bit more advanced:
+ * int response = hc.getConnection().getResponseCode();
+ * </pre>
+ *
+ * @see java.net.HttpURLConnection
```

Comment:
Make a note here that instances of the class shouldn't be shared across threads.